zkevm-circuits/src/evm_circuit/execution/logs.rs

```diff
@@ -0,0 +1,614 @@
+    },
+    util::Expr,
+};
+use array_init::array_init;
+use eth_types::Field;
+use eth_types::{evm_types::GasCost, evm_types::OpcodeId, ToLittleEndian, ToScalar};
+use halo2_proofs::{circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct LogGadget<F> {
+    same_context: SameContextGadget<F>,
+    // memory address
+    memory_address: MemoryAddressGadget<F>,
+    topics: [Cell<F>; 4],
+    topic_selectors: [Cell<F>; 4],
+
+    contract_address: Cell<F>,
+    is_static_call: Cell<F>,
+    is_persistent: Cell<F>,
+    tx_id: Cell<F>,
+    memory_expansion: MemoryExpansionGadget<F, 1, N_BYTES_MEMORY_WORD_SIZE>,
+}
+
+impl<F: Field> ExecutionGadget<F> for LogGadget<F> {
+    const NAME: &'static str = "LOG";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::LOG;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        // check opcode is in [log0,log4], already do SameContextGadget::construct
+        let mstart = cb.query_cell();
+        let msize = cb.query_rlc();
+
+        // Pop mstart_address, msize from stack
+        cb.stack_pop(mstart.clone().expr());
+        cb.stack_pop(msize.expr());
+
+        // read tx id
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+
+        // constrain not in static call
+        let is_static_call = cb.call_context(None, CallContextFieldTag::IsStatic);
+        cb.require_zero("is_static_call is false", is_static_call.expr());
+
+        // check contract_address in CallContext & TxLog
+        // use call context's  callee address as contract address
+        let contract_address = cb.call_context(None, CallContextFieldTag::CalleeAddress);
+        let is_persistent = cb.call_context(None, CallContextFieldTag::IsPersistent);
+        cb.condition(is_persistent.expr(), |cb| {
+            cb.tx_log_lookup(
+                tx_id.expr(),
+                TxLogFieldTag::Address,
+                0.expr(),
+                contract_address.expr(),
+            );
+        });
+
+        // constrain topics in logs
+        let topics = array_init(|_| cb.query_cell());
+        let topic_selectors: [Cell<F>; 4] = array_init(|_| cb.query_cell());
+        for (idx, topic) in topics.iter().enumerate() {
+            cb.condition(topic_selectors[idx].expr(), |cb| {
+                cb.stack_pop(topic.expr());
+            });
+            cb.condition(topic_selectors[idx].expr() * is_persistent.expr(), |cb| {
+                cb.tx_log_lookup(tx_id.expr(), TxLogFieldTag::Topic, idx.expr(), topic.expr());
+            });
+        }
+
+        let opcode = cb.query_cell();
+        let topic_count = opcode.expr() - OpcodeId::LOG0.as_u8().expr();
+        let selector_exprs = topic_selectors.clone().map(|idx| idx.expr());
+
+        // TOPIC_COUNT == Non zero topic count
+        cb.require_equal(
+            " sum of topic selectors = topic_count ",
+            topic_count.clone(),
+            sum::expr(selector_exprs),
```

Comment:
```suggestion
            sum::expr(&topic_selectors),
```

Reply:
updated in [77fe7a5](https://github.com/appliedzkp/zkevm-circuits/pull/335/commits/77fe7a57618c1a13a4cfee7a439daceedc2d4cdf) thanks